pennylane/compiler/compiler.py

```diff
@@ -17,7 +17,14 @@
 from importlib import reload
 from collections import defaultdict
 import dataclasses
-import pkg_resources
+import sys
+
+if sys.version_info[:2] >= (3, 10):
+    import importlib.metadata as importlib_metadata
```

Comment:
@timmysilv What do you think of using `import.metadata` instead of the 3rd party `importlib_metadata` for Python >= 3.10 ?

Reply:
I figure it's easier to just use `importlib_metadata` always, because it's going to have to be a dependency anyways and it works with all versions. once we drop 3.9, we'll use the standard `importlib.metadata` everywhere. do you have strong motivations for using different packages in different versions?

Reply:
@timmysilv I'm in favor of using `importlib.metadata` wherever possible to reduce the dependency on 3rd-party code (as you mentioned, after dropping 3.9, we don't need to install `importlib_metadata`) and the boilerplate wrapper methods added by `importlib_metadata` around `entry_points`. I'm not sure if this approach is widely practiced in PL, and as such, I raised this argument to learn more about it

Reply:
i agree that we don't want additional dependencies, but the below line will fail if the user does not have `importlib_metadata` installed (which will probably happen often if it's not a dependency). I think our options are to add the dependency, or dispatch manually [like I originally did](https://github.com/PennyLaneAI/pennylane/pull/4783/commits/7bbf201748d394d58bc6055ee6650377fe950aa2) in my PR. I'd prefer the dependency because it's a pure-python package written by python themselves and it's only [a few files](https://github.com/python/importlib_metadata/tree/main/importlib_metadata), but I'm fine with the other option. Alternatively, we can just wait until we drop 3.9, but that might be too far away (if we add 3.12 first, we have to add `setuptools` as a dependency anyway so this would all be for naught)

Reply:
Thanks @timmysilv!

Reply:
There's a corner case with the dispatch approach. Let's use `defaultdict` to mask `KeyError` exceptions:

``` python
entries = (
    defaultdict(dict, metadata.entry_points())["pennylane.plugins"]
    if version_info[:2] == (3, 9)
    # pylint:disable=unexpected-keyword-arg
    else metadata.entry_points(group="pennylane.plugins")
)
```